### PR TITLE
feat: Add the stdlib_diff tool to compare gno and go standard libraries

### PR DESCRIPTION
**Pull request by FloRichardAloeCorp**
As requested by the issue [https://github.com/gnolang/gno/issues/1310](#1310),  the aim of this pull request is to add a tool under the misc directory to compare the Gno and Go standard libraries.

The tool generates a differences report per package.

To test the tool, build it, and refer to the README for usage instructions. The tool can be used in both ways: comparing Go to Gno or comparing Gno to Go.

I'm questioning the relevance of adding unit tests to this tool. If you believe it needs some, just let me know!

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [x ] Provided any useful hints for running manual tests
- [x ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

**Review by thehowl — COMMENTED** [review]
Thank you for your contribution!

Sorry for the wait, this is an initial to improve. Nice work, but you can improve.

Some functionality missing on the generated report:

1. Formatting is still very funky. Use CSS to render tabs correctly; they should render as 4 spaces.

image

2. Ideally, for the added/deleted lines, make the `+` and `-` symbols non selectable. This makes it so that if I select the lines in either side of the page I can copy and paste the code without copying the symbols.
3. I believe the report's performance could be improved. Consider using the library I mention in the comments (diffmatchpatch) and parallelizing execution. You can parallelize execution by having a goroutine trying to first of all find all of the directories in the source and destination, and then sending off to a channel individual package names.
4. The tool does not currently perform diffing recursively. Ie. creating the report on my stdlibs it does not generate a diff for package `regexp/syntax`. Consider searching using something like `filepath.WalkDir`.
5. On the report, please make the lines which are == greyed out (putting css `opacity: 0.75` probably works)

File: misc/stdlib_diff/diffstatus.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package main
+
+type diffStatus uint
+
+const (
+	MISSING_IN_SRC diffStatus = 1
+	MISSING_IN_DST diffStatus = 2
+	HAS_DIFF       diffStatus = 3
+	NO_DIFF        diffStatus = 4
```
[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/diffstatus.go]: Fixed In:

[6d9af8f](https://github.com/FloRichardAloeCorp/gno/commit/6d9af8f2aba631ed86283ad667441c5e13c777a5)

File: misc/stdlib_diff/report.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+func NewReportBuilder(srcPath, dstPath, outDir string, srcIsGno bool) (*ReportBuilder, error) {
+	packageTemplate, err := template.ParseFiles("templates/package_diff_template.html")
+	if err != nil {
+		return nil, err
+	}
+
+	indexTemplate, err := template.ParseFiles("templates/index_template.html")
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/report.go]: Fixed in:

[7c7a7ab](https://github.com/FloRichardAloeCorp/gno/commit/7c7a7ab8f5b99db95f58e87f4b26ae5d51c1ae60)

File: misc/stdlib_diff/filediff.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+		Dst:           dst,
+		DiffAlgorithm: diffAlgorithm,
+	}, nil
+}
+
+// Differences returns the differences in lines between the source and
+// destination files using the configured diff algorithm.
+func (f *FileDiff) Differences() ([]LineDifferrence, []LineDifferrence) {
+	srcDiff, dstDiff := f.DiffAlgorithm.Do()
+	return srcDiff, dstDiff
```
[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/filediff.go]: Fixed in:

[ff6f98f](https://github.com/FloRichardAloeCorp/gno/commit/ff6f98fb19d34bc3e3ee00d6abc560eceff04ff2)

File: misc/stdlib_diff/filediff.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	return &FileDiff{
+		Src:           src,
+		Dst:           dst,
+		DiffAlgorithm: diffAlgorithm,
+	}, nil
+}
+
+// Differences returns the differences in lines between the source and
+// destination files using the configured diff algorithm.
+func (f *FileDiff) Differences() ([]LineDifferrence, []LineDifferrence) {
```
[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/filediff.go]: Fixed in:

[019016f](https://github.com/FloRichardAloeCorp/gno/commit/019016f24361935681ed33d4d84d8d4d4eb9e43c)

File: misc/stdlib_diff/myers.go
[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/myers.go:1]: I missed this library ! I will integrate it in the tool.

[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/myers.go:1]: I've checked and tested the library (v1.3.1), it seems that the current version is giving wrong results. I've tested with an older version (v.1.1.0) and it seems to work but I'm not convinced about the stability. I think that it is better to keep the actual myers implementation for now  and if when the lib is fixed we switch to it.

File: misc/stdlib_diff/algorithm.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+func AlgorithmFactory(src, dst []string, algoType string) (Algorithm, error) {
+	switch algoType {
+	case MYERS:
+		return NewMyers(src, dst), nil
+	default:
+		return nil, errors.New("unknown algorithm type")
+	}
+}
```
[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/algorithm.go]: Yes I agree, fixed in:

[95df663](https://github.com/FloRichardAloeCorp/gno/commit/95df66316135f9f5002bb59bdf5642cb73453b1c)

File: misc/stdlib_diff/filediff.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import (
+	"bufio"
+	"os"
+)
+
+// FileDiff is a struct for comparing differences between two files.
+type FileDiff struct {
+	Src           []string  // Lines of the source file.
+	Dst           []string  // Lines of the destination file.
+	DiffAlgorithm Algorithm // Algorithm used for comparison.
```
[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/filediff.go]: Fixed in:

[95df663](https://github.com/FloRichardAloeCorp/gno/commit/95df66316135f9f5002bb59bdf5642cb73453b1c)

File: misc/stdlib_diff/algorithm.go [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package main
+
+import "errors"
+
+const (
+	MYERS = "myers"
+)
+
+type Algorithm interface {
+	Do() (srcDiff []LineDifferrence, dstDiff []LineDifferrence)
```
[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/algorithm.go]: Fixed in:

[95df663](https://github.com/FloRichardAloeCorp/gno/commit/95df66316135f9f5002bb59bdf5642cb73453b1c)

File: misc/stdlib_diff/myers.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+// Do performs the Myers algorithm to find the differences between source and destination files.
+// It returns the differences as two slices of LineDifferrence representing source and destination changes.
+func (m *Myers) Do() ([]LineDifferrence, []LineDifferrence) {
+	operations := m.doMyers()
+
+	srcIndex, dstIndex, insertCount, deleteCount := 0, 0, 0, 0
+	dstDiff := make([]LineDifferrence, 0)
+	srcDiff := make([]LineDifferrence, 0)
```
[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/myers.go]: Fixed in:

[95df663](https://github.com/FloRichardAloeCorp/gno/commit/95df66316135f9f5002bb59bdf5642cb73453b1c)

File: misc/stdlib_diff/myers.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+	for _, op := range operations {
+		switch op {
+		case INSERT:
+			dstDiff = append(dstDiff, LineDifferrence{Line: "+" + m.dst[dstIndex], Operation: op.String()})
+			srcDiff = append(srcDiff, LineDifferrence{Line: "", Operation: MOVE.String()})
+			dstIndex += 1
```
[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/myers.go]: Fixed in:

[95df663](https://github.com/FloRichardAloeCorp/gno/commit/95df66316135f9f5002bb59bdf5642cb73453b1c)

File: misc/stdlib_diff/filediff.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+// LineDifferrence represents a difference in a line during file comparison.
+type LineDifferrence struct {
+	Line      string // The line content.
+	Operation string // The operation performed on the line (e.g., "add", "delete", "equal").
```
[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/filediff.go]: Fixed in :

[5ffff78](https://github.com/FloRichardAloeCorp/gno/commit/5ffff7878c4eca6384e22471fbd62e9cc15fa108)

I've also cleaned operation.go in this commit.

File: misc/stdlib_diff/filediff.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+// Differences returns the differences in lines between the source and
+// destination files using the configured diff algorithm.
+func (f *FileDiff) Differences() ([]LineDifferrence, []LineDifferrence) {
+	srcDiff, dstDiff := f.DiffAlgorithm.Do()
+	return srcDiff, dstDiff
+}
+
+// getFileLines reads and returns the lines of a file given its path.
+func getFileLines(p string) []string {
```
[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/filediff.go]: Fixed in:

[b6b3ad4](https://github.com/FloRichardAloeCorp/gno/commit/b6b3ad4a6824db81c18cdf8ad042366959a66e0e)

File: misc/stdlib_diff/filediff.go [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		lines = append(lines, scanner.Text())
+	}
+
+	if err := scanner.Err(); err != nil {
+		return lines
+	}
+
+	return lines
```
[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/filediff.go]: Fixed in:

[b6b3ad4](https://github.com/FloRichardAloeCorp/gno/commit/b6b3ad4a6824db81c18cdf8ad042366959a66e0e)

File: misc/stdlib_diff/main.go [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+func main() {
+	var srcPath string
+	var dstPath string
+	var outDirectory string
+	var srcIsGno bool
+
+	flag.StringVar(&srcPath, "src", "", "Directory containing packages that will be compared to destination")
+	flag.StringVar(&dstPath, "dst", "", "Directory containing packages; used to compare src packages")
+	flag.StringVar(&outDirectory, "out", "", "Directory where the report will be created")
+	flag.BoolVar(&srcIsGno, "src_is_gno", false, "If true, indicates that the src parameter corresponds to the gno standard libraries")
```
[the remark this answers is not in the document]
Comment (FloRichardAloeCorp) [reply misc/stdlib_diff/main.go:17]: @thehowl I need a clarification on this. Do you want to compare all files ? I mean even files that are not .go or .gno ?

**Comment by thehowl** [comment]
@FloRichardAloeCorp I see some comments are still pending. Would you like me to do a second round of review or are you still making changes?

Thanks!

File: misc/stdlib_diff/myers.go
[the remark this answers is not in the document]
Comment (thehowl) [reply misc/stdlib_diff/myers.go:1]: >it seems that the current version is giving wrong results

Can you specify what you mean?

My concern is about performance. It doesn't need to be spectacular but I'd want the tool to be able to make a full, recursive analysis of the stdlibs directories in < 2min on the machines we test on.

If you're able to do it with your code, it doesn't matter and I'm fine either way

**Comment by FloRichardAloeCorp** [comment]
@thehowl I still have work to do, I was waiting for your answers. I ping you back when I have finished.

**Comment by thehowl** [comment]
Being continued in #2869; closing.

**Pull request closed without merging**